api/client/client.go

```diff
@@ -0,0 +1,254 @@
+	return err
+}
+
+// ModelDelete deletes data
+func (c *Client) ModelDelete(storeID, modelName string, entityIDs ...string) error {
+	req := &pb.ModelDeleteRequest{
+		StoreID:   storeID,
+		ModelName: modelName,
+		EntityIDs: entityIDs,
+	}
+	_, err := c.client.ModelDelete(c.ctx, req)
+	return err
+}
+
+// ModelHas checks if the specified entities exist
+func (c *Client) ModelHas(storeID, modelName string, entityIDs ...string) (bool, error) {
+	req := &pb.ModelHasRequest{
+		StoreID:   storeID,
+		ModelName: modelName,
+		EntityIDs: entityIDs,
+	}
+	resp, err := c.client.ModelHas(c.ctx, req)
+	if err != nil {
+		return false, err
+	}
+	return resp.GetExists(), nil
+}
+
+// ModelFind finds records by query
+func (c *Client) ModelFind(storeID, modelName string, query es.JSONQuery) ([][]byte, error) {
+	queryBytes, err := json.Marshal(query)
+	if err != nil {
+		return [][]byte{}, err
+	}
+	req := &pb.ModelFindRequest{
+		StoreID:   storeID,
+		ModelName: modelName,
+		QueryJSON: queryBytes,
+	}
+	resp, err := c.client.ModelFind(c.ctx, req)
+	if err != nil {
+		return [][]byte{}, err
+	}
+	return resp.GetEntities(), nil
+}
+
+// ModelFindByID finds a record by id
+func (c *Client) ModelFindByID(storeID, modelName, entityID string, entity interface{}) error {
+	req := &pb.ModelFindByIDRequest{
+		StoreID:   storeID,
+		ModelName: modelName,
+		EntityID:  entityID,
+	}
+	resp, err := c.client.ModelFindByID(c.ctx, req)
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal([]byte(resp.GetEntity()), entity)
+	return err
+}
+
+// ReadTransaction returns a read transaction that can be started and used and ended
+func (c *Client) ReadTransaction(storeID, modelName string) (*ReadTransaction, error) {
+	client, err := c.client.ReadTransaction(c.ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &ReadTransaction{client: client, storeID: storeID, modelName: modelName}, nil
+}
+
+// WriteTransaction returns a read transaction that can be started and used and ended
+func (c *Client) WriteTransaction(storeID, modelName string) (*WriteTransaction, error) {
+	client, err := c.client.WriteTransaction(c.ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &WriteTransaction{client: client, storeID: storeID, modelName: modelName}, nil
+}
+
+// Listen provides an update whenever the specified model is updated
+func (c *Client) Listen(storeID, modelName, entityID string) (chan ListenEvent, func(), error) {
+	channel := make(chan ListenEvent)
+	ctx, cancel := context.WithCancel(c.ctx)
+	go func() {
+		defer close(channel)
+		req := &pb.ListenRequest{
+			StoreID:   storeID,
+			ModelName: modelName,
+			EntityID:  entityID,
+		}
+		stream, err := c.client.Listen(ctx, req)
+		if err != nil {
+			channel <- ListenEvent{err: err}
+			return
+		}
+		for {
+			event, err := stream.Recv()
+			if err != nil {
+				status := status.Convert(err)
+				if status == nil || (status != nil && status.Code() != codes.Canceled) {
+					channel <- ListenEvent{err: err}
```

Comment:
If we get an error from the server stream, send it over the channel and close the channel as long as the error isn't the result of canceling the listener.